pandas/_libs/index.pyx

```diff
@@ -278,7 +281,11 @@ cdef class IndexEngine:
 
         self._ensure_mapping_populated()
         values = np.array(self._get_index_values(), copy=False)
+        values_mask = isnaobj(values)
```

Comment:
i dont think values is necessary object dtype is it?  (if it is, should make declaration more specific)

Reply:
@jbrockmendel - you're correct, values is not necessarily an object dtype. in my tests, the following are not object dtype:

1)`DatetimeIndex(['2020-08-05', 'NaT', 'NaT'], dtype='datetime64[ns]', freq=None)` -> int64
2) `array([nan, -1, nan], dtype=object)` -> float64

Knowing this, should anything be changed from your perspective?

Reply:
we'd want to use a dtype-specific `isna` function.  usually that would be pd.isna, but in this file we try pretty hard to avoid importing from outside of our cython code.  Before I can make a suggestion on dealing with that, I need to ask about the dt64 case.

> `DatetimeIndex(['2020-08-05', 'NaT', 'NaT'], dtype='datetime64[ns]', freq=None) -> int64`

IIUC you're `values` here would be `np.array([ 1596585600000000000, -9223372036854775808, -9223372036854775808], dtype=np.int64)`.  Then values_mask would be all-False, despite the fact that the index contains NaTs.  Is that intentional?

If so, then wouldnt it make sense, to override (part of) this method for IndexEngine subclasses where values_mask will always be all-False?

Reply:
That was not intentional. I was under the assumption that `isnaobj` would capture the `NaT` in values_mask as True. Now, from what I understand, `isnaobj` only detects NA-like values when the array dtype=object. Is this correct?

 Interestingly, the dt64 case still passed because both values of NaT were the same integer, so keys were properly updated in `d` (dict): `{-9223372036854775808: [1, 2]}`

I probably shouldn't be relying on the above case and instead use a dt64 na checker.

Could you help point out to me which functions I should be using to check NA for dt64 and float64? Are there other dtypes I should check?

Reply:
> Now, from what I understand, isnaobj only detects NA-like values when the array dtype=object. Is this correct?

Not quite.  If you pass a ndarray[float64] it will correctly identify NaNs, same with ndarray[datetime64ns] and NaTs.  In this case you're just passing int64, which isnaobj will always come give all-False.

> Could you help point out to me which functions I should be using to check NA for dt64 and float64? Are there other dtypes I should check?

IIUC then the check you are implementing is only relevant if `values` can contain NaN, i.e. only floating or complex dtypes.  So I'd suggest 1) implementing an override that skips the new checks for other dtypes and 2) just using `np.isnan(values)`